File: vscode-extension/src/parser/scriptTemplates.js
// Port of core/script/scriptTemplates.cpp — `@stencil` definitions and their expansion.
import { didYouMean, makeDiag, tokenOfStmt } from './scriptDiagnostics.js';
import { MAX_OPS, MAX_TEMPLATE_DEPTH, isStencilUse, unquoteWord } from './scriptTypes.js';

// Name -> index, built once per script: resolveName walks a prefix at a time.
export const templateIndex = (templates) => new Map(templates.map((d, i) => [d.name, i]));

// The call's words, in order, with the literal 'stencil' already dropped.
const callWords = (use) => {
  const words = [];
  let skippedKeyword = false;
  for (const t of use.args) {
    if (t.kind === 'punct') continue;
    if (!skippedKeyword) { skippedKeyword = true; continue; }
    words.push(unquoteWord(t.text));
  }
  return words;
};

const wordCount = (name) => {
  let n = 1;
  for (const c of name) if (c === ' ') n += 1;
  return n;
};

/* Longest defined name that is a prefix of the word run. Only prefixes up to the longest name
 * there is are built, because no longer one can match: building every prefix made one call
 * cost the SQUARE of its word count. */
const resolveName = (words, byName) => {
  let longest = 0;
  for (const name of byName.keys()) longest = Math.max(longest, wordCount(name));

  const limit = Math.min(words.length, longest);
  const prefixes = [];
  let built = '';
  for (let i = 0; i < limit; i += 1) {
    built = built ? `${built} ${words[i]}` : words[i];
    prefixes.push(built);
  }
  for (let n = prefixes.length; n >= 1; n -= 1) {
    const idx = byName.get(prefixes[n - 1]);
    if (idx !== undefined) return { idx, wordsUsed: n };
  }
  return { idx: -1, wordsUsed: 0 };
};

// @1..@n in the body become the call's arguments; every other token passes through.
const substitute = (body, args) => {
  const out = { ...body, args: [] };
  let bad = null;
  for (const t of body.args) {
    if (t.kind !== 'param') {
      out.args.push(t);
      continue;
    }
    const n = parseInt(t.text.slice(1), 10);
    if (!Number.isFinite(n) || n < 1 || n > args.length) {
      bad = t;
      continue;
    }
    out.args.push({ ...t, text: args[n - 1], kind: 'ident' });
  }
  return { stmt: out, bad };
};

/* Expands one `@use stencil <words> [args…]` into the referenced body. Nested uses expand
 * too, capped at MAX_TEMPLATE_DEPTH; a template that reaches itself hits that cap. The
 * statement count is capped as well — the fan-out is what an op cap alone cannot bound. */
export const expandStencilUse = (use, templates, byName, depth, out, diags) => {
  if (depth > MAX_TEMPLATE_DEPTH) {
    diags.push(makeDiag('error', 'E_TEMPLATE_RECURSION', tokenOfStmt(use),
      `templates nest more than ${MAX_TEMPLATE_DEPTH} deep — is one using itself?`));
    return false;
  }

  const words = callWords(use);
  if (words.length === 0) {
    diags.push(makeDiag('error', 'E_ARG_COUNT', tokenOfStmt(use), "'@use stencil' needs a template name"));
    return false;
  }

  const { idx, wordsUsed } = resolveName(words, byName);
  if (idx < 0) {
    const whole = words.join(' ');
    const near = didYouMean(whole, templates.map((d) => d.name));
    diags.push(makeDiag('error', 'E_UNDEFINED_TEMPLATE', tokenOfStmt(use),
      `no template named '${whole}'${near ? ` — did you mean '${near}'?` : ''}`));
    return false;
  }

  const args = words.slice(wordsUsed);
  const { arity, name } = templates[idx];
  if (args.length !== arity) {
    diags.push(makeDiag('error', 'E_TEMPLATE_ARITY', tokenOfStmt(use),
      `template '${name}' takes ${arity} argument(s), got ${args.length}`));
    return false;
  }

  templates[idx].used = true;
  // `substitute` copies each statement, so the body is only ever read here.
  for (const st of templates[idx].body) {
    const { stmt, bad } = substitute(st, args);
    if (bad) {
      diags.push(makeDiag('error', 'E_TEMPLATE_PARAM_INDEX', bad,
        `'${bad.text}' is outside this template's ${arity} argument(s)`));
      return false;
    }
    if (stmt.directive === 'use' && isStencilUse(stmt)) {
      if (!expandStencilUse(stmt, templates, byName, depth + 1, out, diags)) return false;
      continue;
    }
    if (out.length >= MAX_OPS) {
      diags.push(makeDiag('error', 'E_LIMIT_OPS', tokenOfStmt(use), 'the script has too many ops'));
      return false;
    }
    out.push(stmt);
  }
  return true;
};

export const reportUnusedTemplates = (templates, diags) => {
  for (const d of templates) {
    if (d.used) continue;
    diags.push(makeDiag('warning', 'W_UNUSED_TEMPLATE',
      { line: d.line, col: d.col, len: d.len }, `template '${d.name}' is never used`));
  }
};

File: core/script/scriptTemplates.cpp
#include "scriptTemplates.hpp"

#include "scriptArgs.hpp"
#include "scriptDiagnostics.hpp"

#include <cstdlib>

namespace stencil::core::script {

  namespace {

    Token tokenOf(const Stmt& s) {
      Token t;
      t.line = s.line;
      t.col = s.col;
      t.len = s.len;
      t.text = "@" + s.directive;
      return t;
    }

    // The call's words, in order, with `stencil` already dropped.
    std::vector<std::string> callWords(const Stmt& use) {
      std::vector<std::string> words;
      bool skippedKeyword = false;
      for (const Token& t : use.args) {
        if (t.kind == TokenKind::PUNCT) continue;
        if (!skippedKeyword) { skippedKeyword = true; continue; }  // the literal "stencil"
        words.push_back(unquoteWord(t.text));
      }
      return words;
    }

    std::string joinRange(const std::vector<std::string>& w, std::size_t n) {
      std::string out;
      for (std::size_t i = 0; i < n; ++i) {
        if (!out.empty()) out.push_back(' ');
        out += w[i];
      }
      return out;
    }

    // Longest defined name that is a prefix of the word run.
    int resolveName(const std::vector<std::string>& words,
                    const std::vector<TemplateDef>& templates, std::size_t& wordsUsed) {
      int best = -1;
      wordsUsed = 0;
      for (std::size_t n = words.size(); n >= 1; --n) {
        const std::string candidate = joinRange(words, n);
        for (std::size_t k = 0; k < templates.size(); ++k)
          if (templates[k].name == candidate) {
            wordsUsed = n;
            return static_cast<int>(k);
          }
      }
      return best;
    }

    // @1..@n in the body become the call's arguments; every other token passes through.
    Stmt substitute(const Stmt& body, const std::vector<std::string>& args, bool& badIndex,
                    Token& badAt) {
      Stmt out = body;
      out.args.clear();
      for (const Token& t : body.args) {
        if (t.kind != TokenKind::PARAM) {
          out.args.push_back(t);
          continue;
        }
        const int n = std::atoi(t.text.c_str() + 1);
        if (n < 1 || n > static_cast<int>(args.size())) {
          badIndex = true;
          badAt = t;
          continue;
        }
        Token filled = t;
        filled.text = args[static_cast<std::size_t>(n - 1)];
        filled.kind = TokenKind::IDENT;
        out.args.push_back(filled);
      }
      return out;
    }

  }  // namespace

  bool expandStencilUse(const Stmt& use, std::vector<TemplateDef>& templates, int depth,
                        std::vector<Stmt>& out, std::vector<Diagnostic>& diags) {
    if (depth > MAX_TEMPLATE_DEPTH) {
      diags.push_back(makeDiag(Severity::ERROR, "E_TEMPLATE_RECURSION", tokenOf(use),
                               "templates nest more than " + std::to_string(MAX_TEMPLATE_DEPTH) +
                                   " deep — is one using itself?"));
      return false;
    }

    const std::vector<std::string> words = callWords(use);
    if (words.empty()) {
      diags.push_back(makeDiag(Severity::ERROR, "E_ARG_COUNT", tokenOf(use),
                               "'@use stencil' needs a template name"));
      return false;
    }

    std::size_t used = 0;
    const int idx = resolveName(words, templates, used);
    if (idx < 0) {
      std::vector<std::string> names;
      for (const TemplateDef& d : templates) names.push_back(d.name);
      const std::string near = didYouMean(joinRange(words, words.size()), names);
      diags.push_back(makeDiag(Severity::ERROR, "E_UNDEFINED_TEMPLATE", tokenOf(use),
                               "no template named '" + joinRange(words, words.size()) + "'" +
                                   (near.empty() ? "" : " — did you mean '" + near + "'?")));
      return false;
    }

    const std::vector<std::string> args(words.begin() + static_cast<long>(used), words.end());
    const int arity = templates[static_cast<std::size_t>(idx)].arity;
    if (static_cast<int>(args.size()) != arity) {
      diags.push_back(makeDiag(Severity::ERROR, "E_TEMPLATE_ARITY", tokenOf(use),
                               "template '" + templates[static_cast<std::size_t>(idx)].name +
                                   "' takes " + std::to_string(arity) + " argument(s), got " +
                                   std::to_string(args.size())));
      return false;
    }

    templates[static_cast<std::size_t>(idx)].used = true;
    // Copy the body before recursing: expansion may mark other templates used, and a
    // nested @use must not see a half-substituted parent.
    const std::vector<Stmt> body = templates[static_cast<std::size_t>(idx)].body;

    for (const Stmt& st : body) {
      bool badIndex = false;
      Token badAt;
      Stmt filled = substitute(st, args, badIndex, badAt);
      if (badIndex) {
        diags.push_back(makeDiag(Severity::ERROR, "E_TEMPLATE_PARAM_INDEX", badAt,
                                 "'" + badAt.text + "' is outside this template's " +
                                     std::to_string(arity) + " argument(s)"));
        return false;
      }
      if (filled.directive == "use" && !filled.args.empty() &&
          unquoteWord(filled.args[0].text) == "stencil") {
        if (!expandStencilUse(filled, templates, depth + 1, out, diags)) return false;
        continue;
      }
      out.push_back(filled);
    }
    return true;
  }

  void reportUnusedTemplates(const std::vector<TemplateDef>& templates,
                             std::vector<Diagnostic>& diags) {
    for (const TemplateDef& d : templates) {
      if (d.used) continue;
      Token t;
      t.line = d.line;
      t.col = d.col;
      t.len = d.len;
      diags.push_back(makeDiag(Severity::WARNING, "W_UNUSED_TEMPLATE", t,
                               "template '" + d.name + "' is never used"));
    }
  }

}  // namespace stencil::core::script
